### ISSUE TITLE
As is, publishing a Cost Profile will have duplicate information. Should a subclass like costProfileItem be used to simplify the entry and display

### ISSUE THREAD

**Issue by cwd-mparsons** [CredentialEngine/vocabularies#440]
A cost profile with two direct costs would be published as follows. The bolded sections are repeated for each cost profile"

```
"ceterms:estimatedCost": [
		{
			"@type": "ceterms:CostProfile",
			**"ceterms:name": " test cost profile",**
			"ceterms:price": 199.99,
			**"ceterms:currency": "USD",**
			**"ceterms:costDetails": { "@id": "http://example.com?t=costdetails" },
			"ceterms:description": "this description is repeated for each cost item.",**
			"ceterms:audienceType": [
				{
					"@type": "ceterms:CredentialAlignmentObject",
					"ceterms:targetNode": { "@id": "audience:CurrentMilitary" },
					"ceterms:targetNodeName": "Current Military"
				},
				{
					"@type": "ceterms:CredentialAlignmentObject",
					"ceterms:targetNode": { "@id": "audience:CurrentMilitaryDependent" },
					"ceterms:targetNodeName": "Current Military Dependent"
				},
				{
					"@type": "ceterms:CredentialAlignmentObject",
					"ceterms:targetNode": { "@id": "audience:CurrentMilitarySpouse" },
					"ceterms:targetNodeName": "Current Military Spouse"
				}
			],
			**"ceterms:jurisdiction": [
				{
					"@type": "ceterms:JurisdictionProfile",
					"ceterms:globalJurisdiction": true
				}
			],**
			"ceterms:residencyType": [
				{
					"@type": "ceterms:CredentialAlignmentObject",
					"ceterms:targetNode": { "@id": "residency:International" },
					"ceterms:targetNodeName": "Foreign/International"
				},
				{
					"@type": "ceterms:CredentialAlignmentObject",
					"ceterms:targetNode": { "@id": "residency:InState" },
					"ceterms:targetNodeName": "In-State"
				}
			],
			"ceterms:directCostType": [
				{
					"@type": "ceterms:CredentialAlignmentObject",
					"ceterms:targetNode": { "@id": "costType:AggregateCost" },
					"ceterms:targetNodeName": "Aggregated (ALL) Costs",
					"ceterms:targetNodeDescription": "The sum of direct costs"
				}
			]
		},
		{
			"@type": "ceterms:CostProfile",
			**"ceterms:name": " test cost profile",**
			"ceterms:price": 59.99,
			**"ceterms:currency": "USD",
			"ceterms:costDetails": { "@id": "http://example.com?t=costdetails" },
			"ceterms:description": "this description is repeated for each cost item.",**
			"ceterms:audienceType": [
				{
					"@type": "ceterms:CredentialAlignmentObject",
					"ceterms:targetNode": { "@id": "audience:FormerMilitary" },
					"ceterms:targetNodeName": "Former Military / Veteran"
				},
				{
					"@type": "ceterms:CredentialAlignmentObject",
					"ceterms:targetNode": { "@id": "audience:FormerMilitaryDependent" },
					"ceterms:targetNodeName": "Former Military Dependent"
				},
				{
					"@type": "ceterms:CredentialAlignmentObject",
					"ceterms:targetNode": { "@id": "audience:FormerMilitarySpouse" },
					"ceterms:targetNodeName": "Former Military Spouse"
				},
				{
					"@type": "ceterms:CredentialAlignmentObject",
					"ceterms:targetNode": { "@id": "audience:FormerStudent" },
					"ceterms:targetNodeName": "Former Student / Alumni"
				}
			],
			**"ceterms:jurisdiction": [
				{
					"@type": "ceterms:JurisdictionProfile",
					"ceterms:globalJurisdiction": true
				}
			],**
			"ceterms:residencyType": [
				{
					"@type": "ceterms:CredentialAlignmentObject",
					"ceterms:targetNode": { "@id": "residency:OutOfState" },
					"ceterms:targetNodeName": "Out-Of-State"
				},
				{
					"@type": "ceterms:CredentialAlignmentObject",
					"ceterms:targetNode": { "@id": "residency:Regional" },
					"ceterms:targetNodeName": "Regional"
				}
			],
			"ceterms:directCostType": [
				{
					"@type": "ceterms:CredentialAlignmentObject",
					"ceterms:targetNode": { "@id": "costType:EnrollmentFee" },
					"ceterms:targetNodeName": "Enrollment Fee",
					"ceterms:targetNodeDescription": "Cost to enroll in the resource being described."
				}
			]
		}
	],
```
An alternative, using a placeholder property of costProfileItem would appear as follows. The common properties appear once, followed by multiple costProfileItems:
```
	"ceterms:estimatedCost": [
		{
			"@type": "ceterms:CostProfile",
			**"ceterms:name": " test cost profile",
			"ceterms:currency": "USD",
			"ceterms:costDetails": { "@id": "http://example.com?t=costdetails" },
			"ceterms:description": "this description is repeated for each cost item.",
			"ceterms:jurisdiction": [
				{
					"@type": "ceterms:JurisdictionProfile",
					"ceterms:globalJurisdiction": true
				}
			],**
			"ceterms:costItem": [
				{
					"@type": "ceterms:CostProfileItem",
					"ceterms:price": 199.99,
					"ceterms:audienceType": [
						{
							"@type": "ceterms:CredentialAlignmentObject",
							"ceterms:targetNode": { "@id": "audience:CurrentMilitary" },
							"ceterms:targetNodeName": "Current Military"
						},
						{
							"@type": "ceterms:CredentialAlignmentObject",
							"ceterms:targetNode": { "@id": "audience:CurrentMilitaryDependent" },
							"ceterms:targetNodeName": "Current Military Dependent"
						},
						{
							"@type": "ceterms:CredentialAlignmentObject",
							"ceterms:targetNode": { "@id": "audience:CurrentMilitarySpouse" },
							"ceterms:targetNodeName": "Current Military Spouse"
						}
					],

					"ceterms:residencyType": [
						{
							"@type": "ceterms:CredentialAlignmentObject",
							"ceterms:targetNode": { "@id": "residency:International" },
							"ceterms:targetNodeName": "Foreign/International"
						},
						{
							"@type": "ceterms:CredentialAlignmentObject",
							"ceterms:targetNode": { "@id": "residency:InState" },
							"ceterms:targetNodeName": "In-State"
						}
					],
					"ceterms:directCostType": [
						{
							"@type": "ceterms:CredentialAlignmentObject",
							"ceterms:targetNode": { "@id": "costType:AggregateCost" },
							"ceterms:targetNodeName": "Aggregated (ALL) Costs",
							"ceterms:targetNodeDescription": "The sum of direct costs"
						}
					]
				},
				{
					"@type": "ceterms:CostProfileItem",
					"ceterms:price": 59.99,
					"ceterms:audienceType": [
						{
							"@type": "ceterms:CredentialAlignmentObject",
							"ceterms:targetNode": { "@id": "audience:FormerMilitary" },
							"ceterms:targetNodeName": "Former Military / Veteran"
						},
						{
							"@type": "ceterms:CredentialAlignmentObject",
							"ceterms:targetNode": { "@id": "audience:FormerMilitaryDependent" },
							"ceterms:targetNodeName": "Former Military Dependent"
						},
						{
							"@type": "ceterms:CredentialAlignmentObject",
							"ceterms:targetNode": { "@id": "audience:FormerMilitarySpouse" },
							"ceterms:targetNodeName": "Former Military Spouse"
						},
						{
							"@type": "ceterms:CredentialAlignmentObject",
							"ceterms:targetNode": { "@id": "audience:FormerStudent" },
							"ceterms:targetNodeName": "Former Student / Alumni"
						}
					],
					"ceterms:residencyType": [
						{
							"@type": "ceterms:CredentialAlignmentObject",
							"ceterms:targetNode": { "@id": "residency:OutOfState" },
							"ceterms:targetNodeName": "Out-Of-State"
						},
						{
							"@type": "ceterms:CredentialAlignmentObject",
							"ceterms:targetNode": { "@id": "residency:Regional" },
							"ceterms:targetNodeName": "Regional"
						}
					],
					"ceterms:directCostType": [
					{
						"@type": "ceterms:CredentialAlignmentObject",
						"ceterms:targetNode": { "@id": "costType:EnrollmentFee" },
						"ceterms:targetNodeName": "Enrollment Fee",
						"ceterms:targetNodeDescription": "Cost to enroll in the resource being described."
					}
				]
				}
			]
		}
	],
```

**Comment by siuc-nate** [comment]
I agree that the redundant data is annoying and that it's easier for entry to not have to repeat things. 

However, it's impossible to know which properties will be repeated and which will be the same for a given set of cost profiles.  Consider:

- A set in which the cost varies (only) by Jurisdiction
- A set in which the audience type is the same for all costs
- A set in which residency and jurisdiction are the same for all, but audience is different
- A set in which one or more of the cost items has a unique webpage for its information (e.g. for military family members - I believe something like this came up in one of our earlier cost examples)

And so on.  You could possibly consider the currency to be the same across all costs, maybe the description, but then all of your other fields get repeated anyway (e.g., the same audience for each cost profile item). 

What I could see _potentially_ working (or being a headache) is to put the repeatable (cost profile item) properties in both classes, so that the top level class defines a sort of "default" for all of the items, and any individual item can override **only** the relevant property(ies) as needed.  Whether this means letting a CostProfile contain other CostProfiles (which introduces the problem of hierarchies) or creating a CostProfileItem class that is almost identical to a CostProfile class, I'm not sure.  We should probably discuss it.

Something like this:
```
"ceterms:estimatedCost": [
	{
		"@type": "ceterms:CostProfile",
		"ceterms:name": " test cost profile",
		"ceterms:currency": "USD",
		"ceterms:costDetails": { "@id": "http://example.com?t=costdetails" },
		"ceterms:description": "This describes the costs in general",
		"ceterms:jurisdiction": [
			{
				"@type": "ceterms:JurisdictionProfile",
				"ceterms:globalJurisdiction": true
			}
		],
		"ceterms:audienceType": [
			{
				"@type": "ceterms:CredentialAlignmentObject",
				"ceterms:targetNode": { "@id": "audience:CurrentMilitary" },
				"ceterms:targetNodeName": "Current Military"
			},
			{
				"@type": "ceterms:CredentialAlignmentObject",
				"ceterms:targetNode": { "@id": "audience:CurrentMilitaryDependent" },
				"ceterms:targetNodeName": "Current Military Dependent"
			},
			{
				"@type": "ceterms:CredentialAlignmentObject",
				"ceterms:targetNode": { "@id": "audience:CurrentMilitarySpouse" },
				"ceterms:targetNodeName": "Current Military Spouse"
			}
		],
		"ceterms:residencyType": [
			{
				"@type": "ceterms:CredentialAlignmentObject",
				"ceterms:targetNode": { "@id": "residency:International" },
				"ceterms:targetNodeName": "Foreign/International"
			},
			{
				"@type": "ceterms:CredentialAlignmentObject",
				"ceterms:targetNode": { "@id": "residency:InState" },
				"ceterms:targetNodeName": "In-State"
			}
		],
		"ceterms:costItem": [
			{
				"@type": "ceterms:CostProfileItem",
				"ceterms:price": 199.99,
				"ceterms:directCostType": [
					{
						"@type": "ceterms:CredentialAlignmentObject",
						"ceterms:targetNode": { "@id": "costType:AggregateCost" },
						"ceterms:targetNodeName": "Aggregated (ALL) Costs",
						"ceterms:targetNodeDescription": "The sum of direct costs"
					}
				]
			},
			{
				"@type": "ceterms:CostProfileItem",
				"ceterms:price": 50.00,
				"ceterms:directCostType": [
					{
						"@type": "ceterms:CredentialAlignmentObject",
						"ceterms:targetNode": { "@id": "costType:BackgroundCheck" },
						"ceterms:targetNodeName": "Background Check",
						"ceterms:targetNodeDescription": "Cost for a background check."
					}
				],
				"ceterms:jurisdiction": [
					{
						"@type": "ceterms:JurisdictionProfile",
						"ceterms:mainJurisdiction": {
							"@type": "ceterms:GeoCoordinates",
							"ceterms:geoURI": "http://www.geonames.org/4896861/"
						}
					}
				],
			},
			{
				"@type": "ceterms:CostProfileItem",
				"ceterms:price": 59.99,
				"ceterms:audienceType": [
					{
						"@type": "ceterms:CredentialAlignmentObject",
						"ceterms:targetNode": { "@id": "audience:CurrentMilitary" },
						"ceterms:targetNodeName": "Current Military"
					},
					{
						"@type": "ceterms:CredentialAlignmentObject",
						"ceterms:targetNode": { "@id": "audience:CurrentMilitaryDependent" },
						"ceterms:targetNodeName": "Current Military Dependent"
					},
					{
						"@type": "ceterms:CredentialAlignmentObject",
						"ceterms:targetNode": { "@id": "audience:CurrentMilitarySpouse" },
						"ceterms:targetNodeName": "Current Military Spouse"
					},
					{
						"@type": "ceterms:CredentialAlignmentObject",
						"ceterms:targetNode": { "@id": "audience:FormerMilitary" },
						"ceterms:targetNodeName": "Former Military / Veteran"
					},
					{
						"@type": "ceterms:CredentialAlignmentObject",
						"ceterms:targetNode": { "@id": "audience:FormerMilitaryDependent" },
						"ceterms:targetNodeName": "Former Military Dependent"
					},
					{
						"@type": "ceterms:CredentialAlignmentObject",
						"ceterms:targetNode": { "@id": "audience:FormerMilitarySpouse" },
						"ceterms:targetNodeName": "Former Military Spouse"
					},
					{
						"@type": "ceterms:CredentialAlignmentObject",
						"ceterms:targetNode": { "@id": "audience:FormerStudent" },
						"ceterms:targetNodeName": "Former Student / Alumni"
					}
				],
				"ceterms:directCostType": [
					{
						"@type": "ceterms:CredentialAlignmentObject",
						"ceterms:targetNode": { "@id": "costType:EnrollmentFee" },
						"ceterms:targetNodeName": "Enrollment Fee",
						"ceterms:targetNodeDescription": "Cost to enroll in the resource being described."
					}
				]
			}
		]
	}
]
```

But that might actually be _more_ confusing to enter and process...I'm not sure.  I would take the approach that any CostProfileItem property would override (as in, replace the value of - not append to or subtract from) the property in its parent CostProfile. Though this means some duplicate information may exist if the CostProfile applies to CurrentMilitary and FormerMilitary, while the CostProfileItem applies to FormerMilitary and FormerStudent.  Maybe that is an acceptable level of duplication, though.

Another way of describing it:

CostProfile would have these properties (identical to how it is now, with the addition of costItem):
- ceterms:assessmentProfiled (is this property useful? we should get rid of it, I think)
- ceterms:audienceType
- ceterms:condition
- ceterms:costDetails
- ceterms:currency
- ceterms:description
- ceterms:directCostType
- ceterms:endDate
- ceterms:jurisdiction
- ceterms:learningOpportunityProfiled (ditto here - i don't think we really use this)
- ceterms:name
- ceterms:paymentPattern
- ceterms:price
- ceterms:region
- ceterms:residencyType
- ceterms:startDate
- ceterms:taskProfiled (and we should do something with task profile so we can get rid of this)
- ceterms:costItem

CostProfileItem would have these properties (the same as CostProfile, minus the oddball properties highlighted above and minus currency, since currency is reliably the same across a set of CostProfiles):
- ceterms:audienceType
- ceterms:condition
- ceterms:costDetails
- ceterms:description
- ceterms:directCostType
- ceterms:endDate
- ceterms:jurisdiction
- ceterms:name
- ceterms:paymentPattern
- ceterms:price
- ceterms:region
- ceterms:residencyType
- ceterms:startDate

Since defaults are defined by the CostProfile, the CostProfileItem might only contain data for one or two of those properties, but it needs the full gamut in order to avoid the problem pointed out above.

I'm not sure it's a good solution - I personally find it a lot simpler to flatten and duplicate the data as needed (i.e. the way we have it now, in terms of the schema), but I can see the value in a default/override way of doing things.

I dunno. I'm conflicted on this one.

**Comment by stuartasutton** [comment]
@cwd-mparsons, can you point me to an actual resource in WorkIt where this problem might be considered significant?  I'd like to look at some actual data.

**Comment by cwd-mparsons** [comment]
@stuartasutton 
Credential 68 has two cost profiles with 5-7 cost items,so a good example.
Editor:
https://credentialfinder.org/editor/credential/68

Registry
http://lr-staging.learningtapestry.com/ce-registry/envelopes/f9750949-3ed6-4308-b5d5-a67342009fa7

NOTE: I noticed in the latter that a codeNotation property is included, and shouldn't as the code is for the internal code table. I added a Trello card to fix this issue.
